packages/app/src/cli/models/app/loader.ts

```diff
@@ -334,23 +333,32 @@ class AppLoader {
     })
     const configPaths = await path.glob(themeConfigPaths)
 
-    const themeExtensions = configPaths.map(async (configurationPath) => {
+    const extensions = configPaths.map(async (configurationPath) => {
       const directory = path.dirname(configurationPath)
-      const configuration = await this.parseConfigurationFile(ThemeExtensionConfigurationSchema, configurationPath)
-      return {
-        directory,
-        configuration,
-        configurationPath,
-        type: configuration.type,
-        graphQLType: extensionGraphqlId(configuration.type),
-        idEnvironmentVariableName: `SHOPIFY_${string.constantize(path.basename(directory))}_ID`,
-        localIdentifier: path.basename(directory),
+      const fileContent = await file.read(configurationPath)
+      const obj = toml.decode(fileContent)
+      const {type} = TypeSchema.parse(obj)
+      const spec = await specForType(type)
+
+      if (!spec) {
+        this.abortOrReport(
+          output.content`Unknown extension type ${output.token.yellow(type)} in ${output.token.path(
+            configurationPath,
+          )}`,
+          undefined,
+          configurationPath,
+        )
+        return undefined
       }
+
+      const configuration = await this.parseConfigurationFile(spec.schema, configurationPath)
+
+      return new ExtensionInstance(configuration, configurationPath, '', directory, spec, undefined, undefined)
```

Comment:
I've defined a base schema for the toml, and all extensions (theme & ui) extend it adding their own properties, defining their own schema.
Theme app extensions in fact use the base schema without adding anything else.

After that the main difference is with `build` and `dev`, but for the momento I'll keep that out of the `ExtensionInstance` object. 

What I envision is a default `build` method that will apply for all UI current ui extensions, and then `theme` will just override that build method in its spec.

Summarizing, I wanted to avoid having a `ThemeInstance` that would be 90% equal to `ExtensionInstance` :)
It could inherit from it, but from now I see no benefit to do it

Reply:
Thanks for the explanation 👍  Are the two instances 90% equal? I don't have a lot of context on theme extensions, but it suprises me there is so much overlap!